File: modules/api/src/main/scala/org/bigasterisk/api/CrashCulprit.scala
package org.bigasterisk.api

import org.apache.spark.sql.{DataFrame, Row}

/**
 * The record a query was processing when it failed.
 *
 * @param row          the record in flight at the moment of failure
 * @param partitionId  the partition whose task failed
 * @param recordIndex  how many records that task had seen, so the record can be located
 *                     in the input by position as well as by value
 * @param error        the exception's type and message
 *
 * @group crashculprit
 */
case class CulpritRecord(row: Row, partitionId: Int, recordIndex: Long, error: String) {
  override def toString: String =
    s"partition $partitionId, record $recordIndex: $row\n  $error"
}

/**
 * A guard that remembers what a query was working on when it died.
 *
 * @group crashculprit
 */
trait CrashCulprit {

  /** Identifier, unique within the session. */
  def id: String

  /**
   * The instrumented DataFrame. Run this rather than the one passed to
   * [[CrashCulpritSupport.guard]], or nothing is recorded.
   */
  def df: DataFrame

  /**
   * The record in flight when the query failed, if it did.
   *
   * `None` while the query has not failed. Note that a task failing is not the same as
   * the query failing: Spark retries, and a record that kills every attempt is the one
   * reported here.
   *
   * ==Batched Python UDFs==
   * This does not work across a batched Python or Arrow UDF. Such a UDF takes a whole
   * batch to another process before any of it can fail, so by the time the exception
   * comes back the guard has already emitted every row of that batch and the record it
   * remembers is the last of the batch, not the one that failed. Guarding *above* the
   * UDF does not help either: the guard would then never see the failing batch at all.
   *
   * Unlike [[PerfProfile.recordLevel]], this cannot be detected and reported. A profile
   * is compromised by a batched operator *below* it, which is part of the plan it holds
   * and can be inspected; a guard is compromised by one *above* it, which is whatever
   * the caller goes on to build and is not knowable from the guard.
   *
   * Express the failing computation in SQL when you need the record named.
   */
  def culprit: Option[CulpritRecord]

  /**
   * [[culprit]] as a JSON object with the record's fields plus `__partitionId`,
   * `__recordIndex` and `__error`, or `null` when the query has not failed.
   *
   * Provided for language bindings that cannot marshal a Spark `Row` across the process
   * boundary — the PySpark front end reads this.
   */
  def culpritJson: String

  /** Forgets what was recorded, so the guard can be reused for another run. */
  def reset(): Unit
}

/**
 * Crash-culprit determination: when a query dies on bad data, say which record killed
 * it.
 *
 * A failing Spark job reports a stack trace and a task id. Neither tells you which of
 * the billion records being processed was the one the code could not handle, and the
 * usual recourse — bisecting the input by hand — is exactly the work this removes.
 *
 * This is BigDebug's crash-culprit primitive (ICSE 2016). Obtain an instance from
 * [[BigAsterisk.crashCulprit]].
 *
 * {{{
 * val guard = BigAsterisk.crashCulprit(spark).guard(orders)
 * try guard.df.selectExpr("CAST(amount AS INT) / 0").collect()
 * catch { case _: Exception => println(guard.culprit.get) }
 * }}}
 *
 * @group crashculprit
 */
trait CrashCulpritSupport {

  /**
   * Watches the records flowing out of `df`, so that if the query fails the last record
   * seen can be reported.
   *
   * The cost is writing each record into a reused buffer — no allocation per record,
   * and no data moves unless something actually fails.
   */
  def guard(df: DataFrame): CrashCulprit

  /** Every guard created in this session that has not been cleared. */
  def active: Seq[CrashCulprit]

  /** Forgets every guard. Already-instrumented DataFrames keep working. */
  def clear(): Unit
}

File: python/bigasterisk/crashculprit.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Crash-culprit determination: when a query dies on bad data, which record killed it.

A failing Spark job reports a stack trace and a task id. Neither says which of the
billion records being processed was the one the code could not handle, and bisecting the
input by hand is exactly the work this removes.

    import bigasterisk

    guard = bigasterisk.crash_culprit(spark).guard(orders)
    try:
        guard.df.selectExpr("parse_json(payload)").collect()
    except Exception:
        print(guard.culprit)
        # partition 0, record 7: {'oid': 'o8', 'cid': 'c2', 'amount': 99999}
        #   IllegalStateException: cannot handle 99999

This is BigDebug's crash-culprit primitive (ICSE 2016).
"""

import json


def _wrap_dataframe(jdf, spark):
    """Wrap a JVM DataFrame as a PySpark DataFrame.

    PySpark has moved this constructor between releases, so try the session-based form
    first and fall back to the older SQLContext form.
    """
    from pyspark.sql import DataFrame

    try:
        return DataFrame(jdf, spark)
    except TypeError:
        return DataFrame(jdf, spark._wrapped)


class CulpritRecord:
    """The record a query was processing when it failed."""

    def __init__(self, payload):
        self.partition_id = payload.pop("__partitionId")
        self.record_index = payload.pop("__recordIndex")
        self.error = payload.pop("__error")
        self.row = payload

    def __repr__(self):
        return "partition %d, record %d: %r\n  %s" % (
            self.partition_id, self.record_index, self.row, self.error)


class CrashCulprit:
    """A guard that remembers what a query was working on when it died."""

    def __init__(self, jguard, spark):
        self._j = jguard
        self._spark = spark

    @property
    def id(self):
        """Identifier, unique in the session."""
        return self._j.id()

    @property
    def df(self):
        """The instrumented DataFrame. Run this, or nothing is recorded."""
        return _wrap_dataframe(self._j.df(), self._spark)

    @property
    def culprit(self):
        """The record in flight when the query failed, or ``None`` if it has not.

        This does not work across a batched Python or Arrow UDF: the whole batch crosses
        to the worker before any of it can fail, so the record remembered is the last of
        the batch rather than the one that failed. Express the failing computation in SQL
        when you need the record named.
        """
        payload = self._j.culpritJson()
        return CulpritRecord(json.loads(payload)) if payload is not None else None

    def reset(self):
        """Forget what was recorded, so the guard can be reused for another run."""
        self._j.reset()

    def __repr__(self):
        culprit = self.culprit
        return "CrashCulprit(%s, %s)" % (
            self.id, "no failure" if culprit is None else repr(culprit))


class CrashCulpritGuards:
    """Crash-culprit determination for a SparkSession."""

    def __init__(self, spark):
        self._spark = spark
        self._support = spark._jvm.org.bigasterisk.api.BigAsterisk.crashCulprit(
            spark._jsparkSession)

    def guard(self, df):
        """Watch the records flowing out of ``df``, to name the one that kills the query.

        The cost is writing each record into a reused buffer — no allocation per record,
        and nothing moves to the driver unless something actually fails.
        """
        if isinstance(df, str):
            df = self._spark.sql(df)
        return CrashCulprit(self._support.guard(df._jdf), self._spark)

    def active(self):
        """Every guard created and not yet cleared."""
        j = self._support.active()
        return [CrashCulprit(j.apply(i), self._spark) for i in range(j.size())]

    def clear(self):
        """Forget every guard. Already-instrumented DataFrames keep working."""
        self._support.clear()
